controller/k8s/metadata_api.go

```diff
@@ -0,0 +1,256 @@
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/metadata"
+	"k8s.io/client-go/metadata/metadatainformer"
+	"k8s.io/client-go/tools/cache"
+)
+
+// MetadataAPI provides shared metadata informers for some Kubernetes resources
+type MetadataAPI struct {
+	promGauges
+
+	Client          kubernetes.Interface
+	inf             map[APIResource]informers.GenericInformer
+	syncChecks      []cache.InformerSynced
+	sharedInformers metadatainformer.SharedInformerFactory
+}
+
+// InitializeMetadataAPI returns an instance of MetadataAPI with metadata
+// informers for the provided resources
+func InitializeMetadataAPI(kubeConfig string, resources ...APIResource) (*MetadataAPI, error) {
+	config, err := k8s.GetConfig(kubeConfig, "")
+	if err != nil {
+		return nil, fmt.Errorf("error configuring Kubernetes API client: %w", err)
+	}
+
+	k8sClient, err := k8s.NewAPIForConfig(config, "", []string{}, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := metadata.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	api, err := newClusterScopedMetadataAPI(k8sClient, client, resources...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, gauge := range api.gauges {
+		if err := prometheus.Register(gauge); err != nil {
+			log.Warnf("failed to register Prometheus gauge %s: %s", gauge.Desc().String(), err)
+		}
+	}
+	return api, nil
+}
+
+func newClusterScopedMetadataAPI(
+	k8sClient kubernetes.Interface,
+	metadataClient metadata.Interface,
+	resources ...APIResource,
+) (*MetadataAPI, error) {
+	sharedInformers := metadatainformer.NewFilteredSharedInformerFactory(
+		metadataClient,
+		resyncTime,
+		metav1.NamespaceAll,
+		nil,
+	)
+
+	api := &MetadataAPI{
+		Client:          k8sClient,
+		inf:             make(map[APIResource]informers.GenericInformer),
+		syncChecks:      make([]cache.InformerSynced, 0),
+		sharedInformers: sharedInformers,
+	}
+
+	for _, resource := range resources {
+		if err := api.addInformer(resource); err != nil {
+			return nil, err
+		}
+	}
+	return api, nil
+}
+
+// Sync waits for all informers to be synced.
+func (api *MetadataAPI) Sync(stopCh <-chan struct{}) {
+	api.sharedInformers.Start(stopCh)
+
+	waitForCacheSync(api.syncChecks)
+}
+
+func (api *MetadataAPI) getLister(res APIResource) (cache.GenericLister, error) {
+	inf, ok := api.inf[res]
+	if !ok {
+		return nil, fmt.Errorf("metadata informer (%v) not configured", res)
+	}
+
+	return inf.Lister(), nil
+}
+
+// GetCached returns the metadata for the supplied object type and name
+func (api *MetadataAPI) GetCached(res APIResource, name string) (*metav1.PartialObjectMetadata, error) {
```

Comment:
Can you clarify the naming and semantics of this?  My understanding is that this uses the metadata shared informer which should work just like the regular shared informers except that it's metadata only.  Unless there's more to it, I think calling this "cached" in the function name confuses more than it helps.

Reply:
Makes sense, I can't recall why I chose that naming :man_shrugging: